sway-core/src/semantic_analysis/ast_node/declaration/declaration.rs

```diff
@@ -54,14 +54,29 @@ impl TyDecl {
                     ty::TyExpression::error(err, var_decl.name.span(), engines)
                 });
 
+                // TODO: Integers shouldn't be anything special. RHS expressions should be written in
+                //       a way to always use the context provided from the LHS, and if the LHS is
+                //       an integer, RHS should properly unify or type check should fail.
+                //       Remove this special case as a part of the initiative of improving type inference.
                 // Integers are special in the sense that we can't only rely on the type of `body`
                 // to get the type of the variable. The type of the variable *has* to follow
                 // `type_ascription` if `type_ascription` is a concrete integer type that does not
                 // conflict with the type of `body` (i.e. passes the type checking above).
                 let return_type = match &*type_engine.get(type_ascription.type_id) {
                     TypeInfo::UnsignedInteger(_) => type_ascription.type_id,
-                    _ => body.return_type,
+                    _ => match &*type_engine.get(body.return_type) {
```

Comment:
I am finding this part a little bit confused. 

If a variable declaration has type ascription, the `TyVarDecl` below should always use the type ascription of the variable.

And after that, we need to do is unify and check the type ascription with the return of the init expression and give the best error/warning possible, right?

That will also "kill" the special case for `ints` here.

Reply:
Yes you are right. But at this stage I was already overwhelmed with the PR that took way too long exactly because I was tracking issues that were popping up and not limiting the scope of this PR.

Knowing that if I change the current behavior, there will very likely be additional errors popping from the remaining cases of expressions that can be on the RHS but do not comply with the provided `ctx.type_annotation()` but rather use `TypeInfo::Unknown` internally, I've opted for just writing the above TODOs at the moment and let this be solved properly as a part of #5562 step 4. Also, the `TyExpression::type_check()` currently do not return `Err` if the unification fails (#5606), etc.

That's why I fear touching this as a part of this PR could lead to additional effort hard to estimate upfront. And since this PR is actually blocking for #5562 I am still of the opinion to solve this particular case as a part of the #5562 step 4. 

Is this ok for you?

Reply:
Silence implies consent :smile: so I am resolving this remark. It will be done as part of #5562.